istioctl/cmd/revision.go

```diff
@@ -0,0 +1,892 @@
+	apimachinery_schema "k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/duration"
+	"k8s.io/apimachinery/pkg/util/validation"
+)
+
+type revisionArgs struct {
+	manifestsPath string
+	name          string
+	verbose       bool
+	sections      []string
+	output        string
+}
+
+const (
+	IstioOperatorCRSection  = "ISTIO-OPERATOR-CR"
+	ControlPlaneSection     = "CONTROL-PLANE"
+	GatewaysSection         = "GATEWAYS"
+	WebhooksSection         = "MUTATING-WEBHOOKS"
+	NamespaceSummarySection = "NAMESPACE-SUMMARY"
+	PodsSection             = "PODS"
+
+	// TODO: This should be moved to istio/api:label package
+	istioTag = "istio.io/tag"
+
+	jsonFormat   = "json"
+	tableFormat  = "table"
+)
+
+var (
+	validFormats = []string{tableFormat, jsonFormat}
+
+	defaultSections = []string{
+		IstioOperatorCRSection,
+		WebhooksSection,
+		ControlPlaneSection,
+		GatewaysSection,
+	}
+
+	verboseSections = []string{
+		IstioOperatorCRSection,
+		WebhooksSection,
+		ControlPlaneSection,
+		GatewaysSection,
+		NamespaceSummarySection,
+		PodsSection,
+	}
+)
+
+var (
+	istioOperatorGVR = apimachinery_schema.GroupVersionResource{
+		Group:    iopv1alpha1.SchemeGroupVersion.Group,
+		Version:  iopv1alpha1.SchemeGroupVersion.Version,
+		Resource: "istiooperators",
+	}
+
+	revArgs = revisionArgs{}
+)
+
+func revisionCommand() *cobra.Command {
+	revisionCmd := &cobra.Command{
+		Use:     "revision",
+		Short:   "Revision centric view of Istio deployment",
+		Aliases: []string{"rev"},
+	}
+	revisionCmd.PersistentFlags().StringVarP(&revArgs.manifestsPath, "manifests", "d", "", mesh.ManifestsFlagHelpStr)
+	revisionCmd.PersistentFlags().BoolVarP(&revArgs.verbose, "verbose", "v", false, "Enable verbose output")
+	revisionCmd.PersistentFlags().StringVarP(&revArgs.output, "output", "o", "table", "Output format for revision description")
+
+	revisionCmd.AddCommand(revisionListCommand())
+	revisionCmd.AddCommand(revisionDescribeCommand())
+	return revisionCmd
+}
+
+// TODO(su225): Fix example, description - command documentation part
+func revisionDescribeCommand() *cobra.Command {
+	describeCmd := &cobra.Command{
+		Use:     "describe",
+		Example: `    istioctl experimental revision describe <revision>`,
+		Short:   "Show details of a revision - customizations, number of pods pointing to it, istiod, gateways etc",
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return fmt.Errorf("revision must be specified")
+			}
+			if len(args) > 1 {
+				return fmt.Errorf("exactly 1 revision should be specified")
+			}
+			isValidFormat := false
+			for _, f := range validFormats {
+				if revArgs.output == f {
+					isValidFormat = true
+					break
+				}
+			}
+			if !isValidFormat {
+				return fmt.Errorf("unknown format %s. It should be %#v", revArgs.output, validFormats)
+			}
+			if revArgs.name != "<default>" {
+				if errs := validation.IsDNS1123Label(revArgs.name); len(errs) > 0 {
```

Comment:
This validation is failing. See https://github.com/istio/istio/pull/29713#issuecomment-759259941

Reply:
Fixed it. It happened after I moved validation to `PreRun`, but forgot to assign revision to `revArgs`.